pkg/controller/vlan/bridge_vlan_controller.go

```diff
@@ -0,0 +1,226 @@
+	"github.com/rancher/harvester-network-controller/pkg/util"
+)
+
+type BridgeVLANController struct {
+	settingClient harvcontroller.SettingClient
+	settingsCache harvcontroller.SettingCache
+	nadCache      cniv1.NetworkAttachmentDefinitionCache
+	bridge        *bridge.Bridge
+}
+
+const (
+	NetworkSettingName       = "network-setting"
+	BridgeVlanControllerName = "bridge-vlan-controller"
+	BridgeName               = "harvester-br0"
+	BridgeCNIName            = "bridge"
+)
+
+func Register(ctx context.Context, setting harvcontroller.SettingController, nad cniv1.NetworkAttachmentDefinitionController) error {
+	controller := &BridgeVLANController{
+		settingClient: setting,
+		settingsCache: setting.Cache(),
+		nadCache:      nad.Cache(),
+		bridge:        bridge.NewBridge(BridgeName),
+	}
+
+	if err := initNetworkSettings(controller.settingClient); err != nil {
+		return err
+	}
+
+	setting.OnChange(ctx, BridgeVlanControllerName, controller.OnChange)
+	setting.OnRemove(ctx, BridgeVlanControllerName, controller.OnRemove)
+
+	return nil
+}
+
+func (c *BridgeVLANController) OnChange(key string, setting *harvesterv1.Setting) (*harvesterv1.Setting, error) {
+	if setting == nil || setting.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	if setting.Value == "" || key != NetworkSettingName {
+		return setting, nil
+	}
+
+	klog.Infof("harvester network setting configured to: %s", setting.Value)
+	networkSetting, err := DecodeNetworkSettings(setting.Value)
+	if err != nil {
+		return nil, fmt.Errorf("decode failed, error: %w, value: %s", err, setting.Value)
+	}
+
+	if networkSetting.NIC == "" && networkSetting.ConfiguredNIC == "" {
+		klog.Errorf("skip network config, both NIC and configured NIC are empty")
+		return nil, nil
+	}
+
+	settingCopy := setting.DeepCopy()
+	if err := c.configBridgeNetwork(networkSetting); err != nil {
+		harvesterv1.SettingConfigured.False(settingCopy)
+		harvesterv1.SettingConfigured.Reason(settingCopy, fmt.Sprintf("failed to config bridge vlan, error:%s", err.Error()))
+		if _, err := c.settingClient.Update(settingCopy); err != nil {
+			return nil, err
+		}
+		return nil, err
+	}
+
+	networkSetting.ConfiguredNIC = networkSetting.NIC
+	settingCopy.Value, err = EncodeNetworkSettings(networkSetting)
+	if err != nil {
+		return nil, fmt.Errorf("encode network settings failed, error: %w, networksettings: %+v", err, networkSetting)
+	}
+
+	harvesterv1.SettingConfigured.True(settingCopy)
+	harvesterv1.SettingConfigured.Reason(settingCopy, "")
+	// update setting will trigger another reconcile
+	return c.settingClient.Update(settingCopy)
+}
+
+func (c *BridgeVLANController) OnRemove(key string, setting *harvesterv1.Setting) (*harvesterv1.Setting, error) {
+	if setting == nil {
+		return nil, nil
+	}
+
+	if setting.Value == "" || key != NetworkSettingName {
+		return setting, nil
+	}
+
+	klog.Info("harvester network setting has been deleted.")
+	networkSetting, err := DecodeNetworkSettings(setting.Value)
```

Comment:
Can we return early if ConfiguredNIC==NIC?

Reply:
In some case, for example nodes reboot and network configurations are lost, we still need to configure network even if ConfiguredNIC==NIC.